src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java

```diff
@@ -845,6 +861,59 @@ private void causeTrackDeathUpdate(EntityLivingBase entityLivingBase) {
         }
     }
 
+    private EquipmentSlotFabric inventory = new EquipmentSlotFabric(this);
+    private EnumMap<EntityEquipmentSlot, SlotLens<Living, ItemStack>> slotLens = new EnumMap<>(EntityEquipmentSlot.class);
+
+    @Inject(method = "onUpdate", locals = LocalCapture.CAPTURE_FAILHARD,
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/EntityLivingBase;getItemStackFromSlot(Lnet/minecraft/inventory/EntityEquipmentSlot;)Lnet/minecraft/item/ItemStack;"))
+    private void onGetItemStackFromSlot(CallbackInfo ci, int i, EntityEquipmentSlot[] slots, int j, int k,
+                                        EntityEquipmentSlot entityEquipmentSlot, ItemStack before) {
+        ItemStack after = this.getItemStackFromSlot(entityEquipmentSlot);
+        EntityLivingBase entity = EntityUtil.toNative(this);
+        if (!ItemStack.areItemStacksEqual(after, before)) {
+            InventoryAdapter<?, ItemStack> slotAdapter;
+            if (entity instanceof EntityPlayerMP) {
+                SlotLens<IInventory, ItemStack> slotLens;
+                IMixinInventoryPlayer inventory = (IMixinInventoryPlayer) ((EntityPlayerMP) entity).inventory;
+                PlayerInventoryLens inventoryLens = (PlayerInventoryLens) inventory.getRootLens();
+                switch (entityEquipmentSlot) {
+                case OFFHAND:
+                    slotLens = inventoryLens.getOffhandLens();
+                    break;
+                case MAINHAND:
+                    HotbarLens<IInventory, ItemStack> hotbarLens = inventoryLens.getMainLens().getHotbar();
```

Comment:
We might not want an event for the player main-hand here.
All cases I can think of are covered by other events:
ChangeInventoryEvent.Held ; UseItemStackEvent ; ChangeInventoryEvent.Pickup.

In case of the held Slot changing this event will even be wrong because the item in the slot did not actually change. Only the selected Slot did.

Reply:
@Faithcaio There is another special case. None of the event mentioned above will be fired when a mod just change the `ItemStack` on main hand quietly (such as modifying NBTs, etc.), but `ChangeEntityEquipmentEvent` need to take over this change. Maybe we need to think again carefully about what to do.